### PR TITLE
introducing docker #189

### PR DESCRIPTION
**Pull request by vonProteus**
hi
i moved files and made some links
this is the easiest way to supply persistent storage for docker
i don't know php so i didn't want mase with it

if you want i can help you with setting up docker pypeline to automaticly built last master to docker hub

* [x] I am the author of this code or the code is public domain
* [x] I release this code into the public domain

### CODE REVIEW

**Comment by ryangurn** [comment]
@vonProteus I am super excited to see these additions, I will add some comments to the PR.

File: PodcastGenerator/.gitignore [hunk 1]
```diff
@@ -1,7 +1,2 @@
-categories.xml
-config.php
-freebox-content.txt
 rsscache_*
-feed.xml
 themes/testtheme
-buttons.xml
```
Comment (ryangurn) [inline PodcastGenerator/.gitignore:7]: Why are there changes to the .gitignore?

File: PodcastGenerator/appdata/.gitignore [hunk 2]
```diff
@@ -0,0 +1,7 @@
+categories.xml
+config.php
+freebox-content.txt
+rsscache_*
+feed.xml
+themes/testtheme
+buttons.xml
```
Comment (ryangurn) [inline PodcastGenerator/appdata/.gitignore:7]: This is in appdata?

**Comment by ryangurn** [comment]
Is there a reason why you renamed a huge number of files without changing them?

File: podcast-generator-entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash -x
+
+mkdir -p ${APACHE_DOCUMENT_ROOT}/appdata/images
+mkdir -p ${APACHE_DOCUMENT_ROOT}/appdata/media
+mkdir -p ${APACHE_DOCUMENT_ROOT}/appdata/themes/default
+cp -r ${APACHE_DOCUMENT_ROOT}/../theme-default/* ${APACHE_DOCUMENT_ROOT}/appdata/themes/default
+chown -R www-data:www-data ${APACHE_DOCUMENT_ROOT}/appdata
+
+apache2-foreground
```
Comment (ryangurn) [inline podcast-generator-entrypoint.sh:9]: I imagine this creates the default folder structures?

File: docker-compose.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    build:
+      context: ./
+      dockerfile: Dockerfile
+    ports:
+     - target: 80
+       published: 3045
+       protocol: tcp
+       mode: host
+    volumes:
+     - ./PodcastGenerator/appdata:/app/PodcastGenerator/appdata
```
Comment (ryangurn) [inline docker-compose.yaml:13]: Looks nominal

**Review by ryangurn — COMMENTED** [review]
@vonProteus I like this implementation, however it might be more effective to avoid shunting everything into an appdata folder with symbolic links to the actual files. I don't know of a great way of doing that at the moment but ill think about it and comment on the PR conversation area.

File: PodcastGenerator/.gitignore (hunk 1 above)
Comment (vonProteus) [reply PodcastGenerator/.gitignore:7]: those file was moved to appdata and i created links to them in appdata then i ignore files in appdata

File: podcast-generator-entrypoint.sh (hunk 3 above)
Comment (vonProteus) [reply podcast-generator-entrypoint.sh:9]: yes

**Comment by vonProteus** [comment]
> Is there a reason why you renamed a huge number of files without changing them?

i moved `PodcastGenerator/themes/default/* → theme-default/` so i can copy it to appdata on container start

**Comment by vonProteus** [comment]
> @vonProteus I like this implementation, however it might be more effective to avoid shunting everything into an appdata folder with symbolic links to the actual files. I don't know of a great way of doing that at the moment but ill think about it and comment on the PR conversation area.

as i mentioned earlier i didn't mess with php
so the easiest way for me was symbolic link
correct way would be to change application that it would refer to files in appdata

appdata is importent becous it is one folder with all changing files (as far i knowe) and can be easly presist in volume

**Comment by emilengler** [comment]
Hello @vonProteus.
Thanks for this Pull Request, Docker is feature that is being requested for a long time now.
However I have a few questions. Why do we need to rename that much? Is it really necessary?
Even though I never used Docker in a productive environment (doesn't work on OpenBSD) I can't imagine that you would need to rename the files that much. Does it still work for people who want to install PG the traditional way?

My personal suggestion would be that we create another repository inside the PodcastGenerator GitHub organization just for Docker. It'll get updated every time Podcast Generator gets updated.
However we would need someone who is willing to maintain that. Any volunteers?

**Comment by vonProteus** [comment]
@emilengler 
> Why do we need to rename that much? Is it really necessary?

yes if we want to have auto installation of default theme
i move only one folder with ```themes/default```

> Does it still work for people who want to install PG the traditional way?

yes
but this update will be breaking one
people will have to do migration

> My personal suggestion would be that we create another repository inside the PodcastGenerator GitHub organization just for Docker. It'll get updated every time Podcast Generator gets updated.

this would be weird solution for me i think that docker stuff should be in this repository

**Comment by emilengler** [comment]
@vonProteus I think the current patch creates more problems than it solves. We should built Docker around Podcast Generator not vice versa (however I can understand some adjustments are necessary).

My plan would be to have something like this: https://github.com/chambana-net/docker-podcastgen

**Comment by vonProteus** [comment]
ok you don't want my solution
i can live with that

i thought my solution wasn't to intrusive

but in this (https://github.com/chambana-net/docker-podcastgen) case you install Podcast Generator in docker container 

in my solution you run Podcast Generator in docker container

**Comment by RicardoVelaC** [comment]
I think the best solution is to support Docker from the main repository. Docker is a standard plus it's an easy and clean way to install PG

**Comment by ryangurn** [comment]
The difficulty with docker is that it is a standard now, but PG was developed before it was one. That being said it makes it slightly more complex to shift the codebase and account for others who are not using docker if we continue with the assumption that everyone will want and know how to use docker. 

The code itself introduces many changes to the overall file hierarchy and the main goal is to ensure we do not break the program for others who are not adopting this technology yet.
> Does it still work for people who want to install PG the traditional way?

Additionally, for many other major applications, there are many different Docker repos that support one piece of software, some of which are included directly into the codebase and others are secondary repositories to keep a separation of codebase and dockerfile. Every organization chooses to set up this structure as they please.
> My personal suggestion would be that we create another repository inside the PodcastGenerator GitHub organization just for Docker. It'll get updated every time Podcast Generator gets updated.

**Comment by emilengler** [comment]
> ok you don't want my solution

Have I ever said that? It's not that I don't want your solution, it's just that I am a bit skeptical about it.

**Comment by emilengler** [comment]
@ryangurn
> Are you open to creating the repo inside the PodcastGenerator Organization?

Yes, I suggested that above. Thats why we habe this organization :)

**Comment by vonProteus** [comment]
if this push request will be marge i will create another one for introducing travis with automatic builds of docker images for arm-v7 arm64-v8 i386 amd64 and depriving them to docker hub like in my repo

**Comment by emilengler** [comment]
Nice. I will have a deeper look on docker and this PR the next days. Also sorry for my late response. I was in holidays

**Comment by ryangurn** [comment]
Personally after reviewing this PR, I still dont see why there is a need to silo all of the code into an appdata area. @vonProteus is there a specific rationale behind changing the folder structure?

**Comment by emilengler** [comment]
If the folder structure is necessary, I think the best compromise to simply create a shell script that manages that.

**Comment by vonProteus** [comment]
consolidate all files that can change in one directory
allow us to expose one volume in docker
which is much easier than exposing bunch of files and directories

**Comment by MrJustreborn** [comment]
What's the state of this PR? - I would like to setup PodcastGenerator using docker.

**Comment by vonProteus** [comment]
@MrJustreborn if you want my docker container it is available at
https://hub.docker.com/r/vonproteus/podcast-generator
and is runnable on linux/386 linux/amd64 linux/arm/v7 linux/arm64
its from my fork with this PR i will stop to maintain it when this PR will be merged and official image will be operational

**Comment by emilengler** [comment]
> @MrJustreborn if you want my docker container it is available at
> 
> https://hub.docker.com/r/vonproteus/podcast-generator
> 
> and is runnable on linux/386 linux/amd64 linux/arm/v7 linux/arm64
> 
> its from my fork with this PR i will stop to maintain it when this PR will be merged and official image will be operational

I will try to find some time next week in order to do some work on it

**Comment by eliliam** [comment]
Was this ever added to the main repo? This is a huge feature that would allow a much larger group of people access and use this tool, not everyone is comfortable working in the terminal especially when it's more than just installing an apt or flatpak package. I'd love to see some work here on Docker compatibility, and according to all the other comments, so would a ton of others! Excited for what the future holds, thanks for the great project!

**Comment by vonProteus** [comment]
i tried to merge it but maintainer was not interested and after 5 months i gave up 
my [image](https://hub.docker.com/r/vonproteus/podcast-generator) still works (i use it) but is out of date

**Pull request closed without merging**